pkg/network/modules.go

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) 2019 Red Hat and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Modules define network function types
+
+// Package network contains all rendered network oriented resources for Nimbess
+package network
+
+import (
+	log "github.com/sirupsen/logrus"
+)
+
+// Gate is a virtual port for a module for traffic to ingress or egress
+type Gate uint64
+
+// PipelineModule type to implement module updates
+type PipelineModule interface {
+	Connect(module PipelineModule, egress bool, gate *Gate) Gate
+	Disconnect(gate Gate, egress bool)
+	GetNextIGate() Gate
+	GetNextEGate() Gate
+	GetName() string
+	SetName(s string)
+	GetIGateMap() map[Gate]PipelineModule
+	SetIGateMap(m map[Gate]PipelineModule)
+	GetEGateMap() map[Gate]PipelineModule
+	SetEGateMap(m map[Gate]PipelineModule)
+	SetMeta(s string)
+	GetMeta() string
+}
+
+// Module represents a generic Network Function.
+// It contains ingress gates and egress gates for traffic flow, which map to the next connected module.
+// Meta used to the name of the Meta Pipeline Module from which this module was created
+type Module struct {
+	Meta   string
+	Name   string
+	IGates map[Gate]PipelineModule
+	EGates map[Gate]PipelineModule
+}
+
+// Switch represents an L2 network switch.
+// Extends Module and defines L2FIB table which holds a map of map to output Gates
+// EGate 999 is always reserved in Switch for connecting to replicator
+type Switch struct {
+	Module
+	L2FIB map[string]Gate
+}
+
+// Port represents a network interface
+type Port struct {
+	PortName   string
+	Virtual    bool
+	DPDK       bool
```

Comment:
does DPDK here mean this is still a data plane agnostic construct? Just wondering if DPDK implies something about the interfaces used.

Reply:
DPDK indicates if the port should be treated as a userspace DPDK port. That doesn't imply a particular data plane, but it does indicate to a data plane how to create this port. For example in BESS, this field is necessary to know whether or not to tell BESS to create a kernel or virtio user type port.